lib/Sema/CSFix.cpp

```diff
@@ -2151,3 +2152,162 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
+  MissingExplicitExistentialCoercion failure(solution, ErasedResultType,
+                                             getLocator());
+  return failure.diagnose(asNote);
+}
+
+bool AddExplicitExistentialCoercion::isRequired(
+    ConstraintSystem &cs, Type resultTy,
+    llvm::function_ref<Optional<Type>(TypeVariableType *)> findExistentialType,
+    ConstraintLocatorBuilder locator) {
+  using ExistentialTypeFinder =
+      llvm::function_ref<Optional<Type>(TypeVariableType *)>;
+
+  struct CoercionChecker : public TypeWalker {
+    bool RequiresCoercion = false;
+
+    ConstraintSystem &cs;
+    ExistentialTypeFinder GetExistentialType;
+
+    CoercionChecker(ConstraintSystem &cs,
+                    ExistentialTypeFinder getExistentialType)
+        : cs(cs), GetExistentialType(getExistentialType) {}
+
+    Action walkToTypePre(Type componentTy) override {
+      // In cases where result references a member type, we need to check
+      // whether such type would is resolved to concrete or not.
+      if (auto *member = componentTy->getAs<DependentMemberType>()) {
+        auto memberBaseTy = getBaseTypeOfDependentMemberChain(member);
+
+        auto typeVar = memberBaseTy->getAs<TypeVariableType>();
+        if (!typeVar)
+          return Action::SkipChildren;
+
+        // If the base is an opened existential type, let's see whether
+        // erase would produce an existential in this case and if so,
+        // we need to check whether any requirements are going to be lost
+        // in process.
+
+        auto existentialType = GetExistentialType(typeVar);
+        if (!existentialType)
+          return Action::SkipChildren;
+
+        auto erasedMemberTy = typeEraseOpenedExistentialReference(
+            Type(member), *existentialType, typeVar, TypePosition::Covariant);
+
+        // If result is an existential type and the base has `where` clauses
+        // associated with its associated types, the call needs a coercion.
+        if (erasedMemberTy->isExistentialType() &&
+            hasConstrainedAssociatedTypes(member)) {
+          RequiresCoercion = true;
+          return Action::Stop;
+        }
+
+        return Action::SkipChildren;
+      }
+
+      // The case where there is a direct access to opened existential type
+      // e.g. `$T` or `[$T]`.
+      if (auto *typeVar = componentTy->getAs<TypeVariableType>()) {
+        if (auto existentialType = GetExistentialType(typeVar)) {
+          RequiresCoercion |= hasConstrainedAssociatedTypes(
+              (*existentialType)->getExistentialLayout());
+          return RequiresCoercion ? Action::Stop : Action::SkipChildren;
+        }
+      }
+
+      return Action::Continue;
+    }
+
+  private:
+    bool hasConstrainedAssociatedTypes(DependentMemberType *member) {
+      auto *assocType = member->getAssocType();
+
+      // If this member has associated type requirements, we are done.
+      if (assocType->getTrailingWhereClause())
```

Comment:
This is looking at the syntactic construct (the trailing where clause), which is not serialized and may not have complete information. Can we instead look at the generic signature to see if there are any requirements that mention this dependent member type?

Reply:
Ah, I didn't know that this was the case, will fix it before merging this.